core/engine/src/value/mod.rs

```diff
@@ -63,38 +65,79 @@ static TWO_E_63: Lazy<BigInt> = Lazy::new(|| {
     BigInt::from(TWO_E_63)
 });
 
-/// A Javascript value
-#[derive(Finalize, Debug, Clone)]
-pub enum JsValue {
-    /// `null` - A null value, for when a value doesn't exist.
+/// The Inner type of a [`JsValue`]. This is the actual value that the `JsValue` holds.
+/// This is not a public API and should not be used directly.
+///
+/// If you need access to the variant, use [`JsValue::variant`] instead.
+#[derive(Finalize, Debug, Clone, PartialEq)]
+enum InnerValue {
     Null,
-    /// `undefined` - An undefined value, for when a field or index doesn't exist.
     Undefined,
-    /// `boolean` - A `true` / `false` value, for if a certain criteria is met.
     Boolean(bool),
-    /// `String` - A UTF-16 string, such as `"Hello, world"`.
-    String(JsString),
-    /// `Number` - A 64-bit floating point number, such as `3.1415`
-    Rational(f64),
-    /// `Number` - A 32-bit integer, such as `42`.
-    Integer(i32),
-    /// `BigInt` - holds any arbitrary large signed integer.
+    Float64(f64),
+    Integer32(i32),
     BigInt(JsBigInt),
-    /// `Object` - An object, such as `Math`, represented by a binary tree of string keys to Javascript values.
-    Object(JsObject),
-    /// `Symbol` - A Symbol Primitive type.
+    String(JsString),
     Symbol(JsSymbol),
+    Object(JsObject),
 }
 
-unsafe impl Trace for JsValue {
+#[allow(unsafe_op_in_unsafe_fn)]
+unsafe impl Trace for InnerValue {
     custom_trace! {this, mark, {
         if let Self::Object(o) = this {
             mark(o);
         }
     }}
 }
 
+/// A Javascript value
+#[derive(Finalize, Debug, Clone, Trace)]
+pub struct JsValue {
+    inner: InnerValue,
+}
+
 impl JsValue {
+    /// The integer zero as a [`JsValue`] constant, for convenience.
+    pub const ZERO: Self = Self {
+        inner: InnerValue::Integer32(0),
+    };
+
+    /// The integer one as a [`JsValue`] constant, for convenience.
+    pub const ONE: Self = Self {
+        inner: InnerValue::Integer32(1),
+    };
+
+    /// `NaN` as a [`JsValue`] constant, for convenience.
+    pub const NAN: Self = Self {
+        inner: InnerValue::Float64(f64::NAN),
+    };
+
+    /// Positive infinity as a [`JsValue`] constant, for convenience.
+    pub const POSITIVE_INFINITY: Self = Self {
+        inner: InnerValue::Float64(f64::INFINITY),
+    };
+
+    /// Negative infinity as a [`JsValue`] constant, for convenience.
+    pub const NEGATIVE_INFINITY: Self = Self {
+        inner: InnerValue::Float64(f64::NEG_INFINITY),
+    };
+
+    /// Undefined as a [`JsValue`] constant, for convenience.
+    pub const UNDEFINED: Self = Self {
+        inner: InnerValue::Undefined,
+    };
+
+    /// Null as a [`JsValue`] constant, for convenience.
+    pub const NULL: Self = Self {
+        inner: InnerValue::Null,
+    };
```

Comment:
Since we have `pub const` functions for all of these values, I would be for removing these values. My reasoning is that our API surface gets smaller and we do not run into the issue of having to break these `const`s, should we change our implementation.

Another reason for both not having these `pub` and not using them internally is that there should be only one way to do things. In my experience that makes rafactoring and assessing the impact / usage of code much easier.

Reply:
I was trying to avoid using `const UNDEFINED: JsValue = JsValue::undefined()` which I now realize didn't help that much. Removed.